rails newを実行した後、各モデルを作りました

daisakuhazui opened the pull request:
# WHY
モデルを作成し、そのテーブルを定義した。
（deviceも導入）

# WHAT
最初にDBがあったほうが開発しやすいため。

Commit:
Answerモデル作成

ターミナルにrails g model answerとする際、user:references question:referencesとしました。モデル作成時にbelongs_toで関連付けがなされています。

Commit:
ポリモーフィック関連性の追加

picturesテーブルとの関連付けを行いました。

Commit:
Genreモデルの作成

genre_usersテーブルを中間テーブルとしてusersテーブルとの関連付けを行っています。

Commit:
Genre_userモデルの作成

genre_usersテーブル（中間テーブル）のためにモデルを作成しました。

Commit:
Pictureモデルの作成

ポリモーフィック関連付けでquestionsテーブルとanswersテーブルの両方に紐付けています。

Commit:
Questionモデルの作成

usersテーブルと関連付けを行ったほか、picturesテーブルとポリモーフィック関連付けを行っています。

Commit:
Userモデルの作成

deviseを使ってuserモデルを作成しました。また、genre_usersテーブルを通してgenresテーブルと関連付けを行っています。

Commit:
YAMLファイル

各モデルの作成時に自動生成。

Commit:
gemの追加・編集

sqlite3を削除してmysql2を追加。deviseを追加。
※rails newを実行する際におそらく-d mysqlを入力していなかったため。

Commit:
テストファイル

各モデル作成時に自動で生成。

Commit:
使用するdatabaseの変更

sqliteを使用する設定になっていたため、mysql2に変更しました。

Commit:
Userモデルのマイグレーションファイルの作成とオプション追加

alias_name, name,にnullを許容しないオプションを追加しました。
また、alias_nameとnameにインデックスを貼ってあります。

Commit:
Genreモデルのマイグレーションを作成

自動生成された後、何も編集していません。

Commit:
Genre_userモデルのマイグレーションファイルを作成

自動生成された状態のままです。

Commit:
Questionモデルのマイグレーションファイルを作成

自動生成された状態のままです。

Commit:
Answerモデルのマイグレーションファイルを作成

自動生成された状態のままです。

Commit:
Pctureモデルのマイグレーションファイルを作成

自動生成後、polymorphic: trueオプションを追加しました。

Commit:
deviseの導入に必須の入力項目を追加

deviseのインストール時に指示されるコードを追加しました。

Commit:
ターミナルでrails g devise:viewsの実行により生成されたファイル

これから作成するアプリケーションではdeviseの画面に変更を加えるため、上記コマンドを実行しました。

Commit:
rails g devise:installによりプロジェクトに追加されたファイル

Commit:
スキーマ情報

rake db:migrateによって毎回更新されるため常にコミットしておく。Railsは最終的に、このファイル内のスキーマ情報を使用してデータベースを扱う。

Commit:
routes/rbにrootパスを追加

ArakawaShintaro commented:
そもそもなんですが、１機能ずつの方がいいですね。
今回だと、device導入とUser以外は無関係です。
PRの粒度が大きすぎて、見るのも大変ですし。。。

ArakawaShintaro commented:
このPRで何をしたかったのか一言で説明できる大きさ、つまり**１機能１ブランチ**でいきましょう！

例えば、今回のPRではモデルとそのテーブルを定義したかったのだと思います。
しかし、deviceの導入までもが混在していますよね。
そうすることで、PRの粒度が大きくなり、レビュー側も見るのが大変になります。

僕なら下記のようにPRを出し、deviceは別PRでやります。

# 何をしたか
今回作成するアプリにおいて、モデルとそのテーブルを先に定義した。
またDBのリレーションも定義した。

# なぜしたか
最初にDBがあった方が開発しやすいため

ArakawaShintaro commented on Gemfile:
```diff
@@ -9,7 +9,7 @@ end
 gem 'rails', '~> 5.1.4'
 # Use sqlite3 as the database for Active Record
-gem 'sqlite3'
+gem 'mysql2'
```
バージョン指定しなくてもエラー出ていませんか？

daisakuhazui replied:
ターミナルでrails sを実行してもエラーは発生しませんでした。

ArakawaShintaro commented on app/models/genre.rb:
```diff
@@ -0,0 +1,4 @@
+class Genre < ApplicationRecord
+  has_many :users, through: :genre_users
+  has_many :genre_users
```
親要素が削除されたら子要素も削除されるようにした方がいいですね。
そういったオプションがあるので調べて見てください！

daisakuhazui replied:
:dependent => :destroyというものがありましたね・・・了解です！

ArakawaShintaro commented on app/models/user.rb:
```diff
@@ -0,0 +1,9 @@
+         :recoverable, :rememberable, :trackable, :validatable
+
+  has_many :genres, through: :genre_users
+  has_many :genre_users
```
親要素が削除されたら子要素も削除されるようにした方がいいですね。
そういったオプションがあるので調べて見てください！

daisakuhazui replied:
こちらも :dependent => :destroyを使います！

ArakawaShintaro commented on config/database.yml:
```diff
@@ -1,25 +1,56 @@
+  host: localhost
+
+
+# # SQLite version 3.x
```
いらないコメントは全て消しましょう。

daisakuhazui replied:
わかりました！

ArakawaShintaro commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
+  devise_for :users
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
```
いらないコメントは消しましょう

daisakuhazui replied:
わかりました！

ArakawaShintaro commented on db/migrate/20171205035152_create_genres.rb:
```diff
@@ -0,0 +1,9 @@
+    create_table :genres do |t|
+      t.string :title
+
+      t.timestamps
```
ジャンルにtimestampが必要なのはなぜでしょう？

daisakuhazui replied:
Questionのように時系列で並べることもないですし不要ですね・・・！

削除します。

ArakawaShintaro replied:
こちらは直さないんですかね？

ArakawaShintaro commented on db/migrate/20171205035347_create_genre_users.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateGenreUsers < ActiveRecord::Migration[5.1]
+  def change
+    create_table :genre_users do |t|
+      t.references :user, foreign_key: true
```
外部キーにはindexを貼りたいですね

ArakawaShintaro replied:
ここはnullを許容してもいいのですか？

daisakuhazui replied:
外部キーとインデックスについて調べてみたところ、下記記事を見つけました。
https://qiita.com/ryouzi/items/2682e7e8a86fd2b1ae47
これによるとreference型を用いると自動でインデックスがはられるようなのですが、正しいですか？
（コードにadd indexを書かなかったのは私が失念していただけなのですが）

また、nullは許容したくないです。null: falseを追加します。

ArakawaShintaro commented on test/fixtures/answers.yml:
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
今回はテストコードを書かないので、test/以下のものはなくて大丈夫です。
なんで削除して起きましょう。
また、これはrails gのときに自動で生成されてしまうので
生成されないようにしてみてください。
ググったらでてきます 👍

ArakawaShintaro replied:
下記のものも同様に全てですね

daisakuhazui replied:
テストコードは書かないのですね。わかりました！

ArakawaShintaro replied:
test 
- rspec

ArakawaShintaro commented:
@daisakuhazui 
もろもろコメントしました。
僕のコメント１つ１つに返答をいただけると！

daisakuhazui committed:
不要なコメントの削除

コメントアウトされていた箇所を削除しました。

ArakawaShintaro commented on Gemfile:
```diff
@@ -32,6 +32,7 @@ gem 'jbuilder', '~> 2.5'
 
 # Use Capistrano for deployment
 # gem 'capistrano-rails', group: :development
+gem 'devise'
```
test

daisakuhazui replied:
test reply

Commit:
テストファイルの削除

今回、テストコードは書かないため、testディレクトリ以下のファイルを全て削除しました。

Commit:
テストファイルが自動生成されないようにする

今回のアプリ開発ではテストコードを書かないため、rails g実行時にテストファイルが自動生成されないようにしました。
また、不要なコメントアウトを削除しました。

Commit:
親要素が削除されたら子要素も削除されるようにする

オプションを追加し、userが削除されればgenre_userも削除されるようにしました。

Commit:
不要なコメントアウトを削除

routes.rbにはこれからコードを追記していくため、不要なコメントを削除しました。

Commit:
dependentオプションのsyntax error修正

" , "が抜けていたため追記しました。

Commit:
null: falseのオプションを追加

外部キー2つに関して、nullを許容しないようにしました。

Commit:
Merge branch 'create_models' of https://github.com/daisakuhazui/original_app into create_models

ArakawaShintaro commented on app/models/user.rb:
```diff
@@ -5,5 +5,5 @@ class User < ApplicationRecord
 
   has_many :genres, through: :genre_users
-  has_many :genre_users
+  has_many :genre_users :dependent => :destroy
```
細かいですが、
```
dependent: :destroy
```
こっちの記述が使われていることが多いかもです

ArakawaShintaro commented on app/models/user.rb:
```diff
@@ -0,0 +1,7 @@
+         :recoverable, :rememberable, :trackable, :validatable
+
+  has_many :genres, through: :genre_users
+  has_many :genre_users, :dependent => :destroy
```
```
dependent: :destroy
```
こちらの記述の方がよくされている気がします

ArakawaShintaro replied:
細かいですが...

daisakuhazui replied:
文字数減るので、そちらの記述でいきますね。

ArakawaShintaro commented:
@daisakuhazui 
timestampが修正されていませんね...
そちらを直したらマージして、次の機能に進みましょう

Commit:
dependentオプションの修正

より一般的な記述に変更しました。コードの文字数が減るのでこちらでいきます！

Commit:
不要なtimestampを削除

DBに不要なカラムが作成されてしまうため、timestampを削除しました。
